purchases/src/main/kotlin/com/revenuecat/purchases/PostReceiptHelper.kt

```diff
@@ -0,0 +1,143 @@
+package com.revenuecat.purchases
+
+import com.revenuecat.purchases.common.AppConfig
+import com.revenuecat.purchases.common.Backend
+import com.revenuecat.purchases.common.BillingAbstract
+import com.revenuecat.purchases.common.PostReceiptDataErrorCallback
+import com.revenuecat.purchases.common.PostReceiptDataSuccessCallback
+import com.revenuecat.purchases.common.ReceiptInfo
+import com.revenuecat.purchases.common.caching.DeviceCache
+import com.revenuecat.purchases.models.StoreProduct
+import com.revenuecat.purchases.models.StoreTransaction
+import com.revenuecat.purchases.subscriberattributes.SubscriberAttributesManager
+import com.revenuecat.purchases.subscriberattributes.getAttributeErrors
+import com.revenuecat.purchases.subscriberattributes.toBackendMap
+
+@Suppress("LongParameterList")
+internal class PostReceiptHelper(
+    private val appConfig: AppConfig,
+    private val backend: Backend,
+    private val billing: BillingAbstract,
+    private val customerInfoHelper: CustomerInfoHelper,
+    private val deviceCache: DeviceCache,
+    private val subscriberAttributesManager: SubscriberAttributesManager
+) {
+    private val finishTransactions: Boolean
+        get() = appConfig.finishTransactions
+
+    /**
+     * This method will post a token and receiptInfo to the backend without consuming any purchases.
+     * It will store that the token was sent to the backend so it doesn't send it again.
+     */
+    fun postTokenWithoutConsuming(
+        purchaseToken: String,
+        storeUserID: String?,
+        receiptInfo: ReceiptInfo,
+        isRestore: Boolean,
+        appUserID: String,
+        marketplace: String?,
+        onSuccess: () -> Unit,
+        onError: (PurchasesError) -> Unit,
+    ) {
+        postReceiptAndSubscriberAttributes(
+            appUserID,
+            purchaseToken,
+            isRestore,
+            receiptInfo,
+            storeUserID,
+            marketplace,
+            onSuccess = { _, _ ->
+                deviceCache.addSuccessfullyPostedToken(purchaseToken)
+                onSuccess()
+            },
+            onError = { error, shouldConsumePurchase, _, _ ->
+                if (shouldConsumePurchase) {
+                    deviceCache.addSuccessfullyPostedToken(purchaseToken)
+                }
+                onError(error)
+            }
+        )
+    }
+
+    /**
+     * This method will post a StoreTransaction and optionally a StoreProduct info to the backend.
+     * It will consume the purchase if finishTransactions is true.
+     */
```

Comment:
should we make clear this one doesn't save the token to the cache?

Reply:
It also does. It's part of the `consumeAndSave` function. Only difference is this method consumes it first